File: src/spirite/pc/ui/Transferables.java
package spirite.pc.ui;

import spirite.base.image_data.Animation;
import spirite.base.image_data.GroupTree.Node;

import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.io.IOException;

public class Transferables {

	
	public static class NodeTransferable implements Transferable {
		public final Node node;
		public NodeTransferable( Node node) {
			this.node = node;
		}
		@Override
		public Object getTransferData(DataFlavor flavor) throws UnsupportedFlavorException, IOException {
			if( flavor.equals(FLAVOR))	return this;
			else throw new UnsupportedFlavorException(flavor);
		}	
		@Override public DataFlavor[] getTransferDataFlavors() {return flavors;}
		@Override public boolean isDataFlavorSupported(DataFlavor flavor) {return flavor.equals(FLAVOR);}

		public static final DataFlavor FLAVOR = new DataFlavor( NodeTransferable.class, "Group Tree Node");
		public static final DataFlavor flavors[] = {FLAVOR};
	}
	
	public static class AnimationTransferable implements Transferable {
		public final Animation animation;
		public AnimationTransferable( Animation animation) {
			this.animation = animation;
		}
		
		@Override
		public Object getTransferData(DataFlavor flavor) throws UnsupportedFlavorException, IOException {
			if( flavor.equals(FLAVOR))	return this;
			else throw new UnsupportedFlavorException(flavor);
		}
		@Override public DataFlavor[] getTransferDataFlavors() {return flavors;}
		@Override public boolean isDataFlavorSupported(DataFlavor flavor) {return flavor.equals(FLAVOR);}

		public static final DataFlavor FLAVOR = new DataFlavor( AnimationTransferable.class, "Spirite Animation");
		public static final DataFlavor flavors[] = {FLAVOR};
	}
}

File: src/main/kotlin/spirite/gui/resources/Transferables.kt
package spirite.gui.resources

import sgui.core.transfer.ITransferObject
import spirite.base.imageData.animation.Animation
import spirite.base.imageData.groupTree.GroupTree.Node
import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.Transferable
import java.awt.datatransfer.UnsupportedFlavorException

object Transferables {
    class NodeTransferable(val node: Node) : Transferable {
        override fun getTransferData(flavor: DataFlavor) =  when(flavor) {
            FLAVOR -> node
            else -> throw UnsupportedFlavorException(flavor)
        }

        override fun getTransferDataFlavors() = flavors
        override fun isDataFlavorSupported(flavor: DataFlavor) = flavor.equals(FLAVOR)

        companion object {
            val FLAVOR = DataFlavor(NodeTransferable::class.java, "Group Tree GroupNode")
            val flavors = arrayOf(FLAVOR)
        }
    }

    class NodeTransferObject(val node: Node) : ITransferObject {
        override val dataTypes: Set<String> get() = setOf(Key)

        override fun getData(type: String) = when(type) {
            Key -> node
            else -> null
        }

        companion object {
            const val Key = "SpiriteNode"
        }
    }

    class AnimationTransferable(val anim: Animation) : Transferable {
        override fun getTransferData(flavor: DataFlavor) =  when(flavor) {
            FLAVOR -> anim
            else -> throw UnsupportedFlavorException(flavor)
        }

        override fun getTransferDataFlavors() = flavors
        override fun isDataFlavorSupported(flavor: DataFlavor) = flavor.equals(FLAVOR)

        companion object {
            val FLAVOR = DataFlavor(AnimationTransferable::class.java, "Spirite Animation")
            val flavors = arrayOf(FLAVOR)
        }
    }
}